src/core/geometry/qgscircularstring.cpp

```diff
@@ -546,6 +546,52 @@ double QgsCircularString::yAt( int index ) const
     return 0.0;
 }
 
+void QgsCircularString::filterVertices( const std::function<bool ( const QgsPoint & )> &filter )
+{
+  bool hasZ = is3D();
+  bool hasM = isMeasure();
+  int size = mX.size();
+
+  double *srcX = mX.data(); // clazy:exclude=detaching-member
+  double *srcY = mY.data(); // clazy:exclude=detaching-member
+  double *srcM = hasM ? mM.data() : nullptr; // clazy:exclude=detaching-member
+  double *srcZ = hasZ ? mZ.data() : nullptr; // clazy:exclude=detaching-member
+
+  double *destX = srcX;
+  double *destY = srcY;
+  double *destM = srcM;
+  double *destZ = srcZ;
+
+  int filteredPoints = 0;
+  for ( int i = 0; i < size; ++i )
+  {
+    double x = *srcX++;
+    double y = *srcY++;
+    double z = hasZ ? *srcZ++ : std::numeric_limits<double>::quiet_NaN();
+    double m = hasM ? *srcM++ : std::numeric_limits<double>::quiet_NaN();
```

Comment:
Just a fun idea, could use a `signalling_NaN` and then check if a floating point exception has been raised to inform about filters that do not check for existence of z/m coordinates. :P